wire-ios-sync-engine/Source/SessionManager/SessionManager.swift

```diff
@@ -785,63 +785,103 @@ public final class SessionManager: NSObject, SessionManagerType {
                 completion(session)
                 onWorkDone()
                 group?.leave()
             } else {
-                let coreDataStack = CoreDataStack(account: account,
-                                                  applicationContainer: self.sharedContainerURL,
-                                                  dispatchGroup: self.dispatchGroup)
+                let coreDataStack = CoreDataStack(
+                    account: account,
+                    applicationContainer: self.sharedContainerURL,
+                    dispatchGroup: self.dispatchGroup
+                )
 
                 if coreDataStack.needsMigration {
                     self.delegate?.sessionManagerWillMigrateAccount(userSessionCanBeTornDown: {})
                 }
 
-                coreDataStack.loadStores { (error) in
+                coreDataStack.loadStores { error in
                     if error != nil {
                         self.delegate?.sessionManagerDidFailToLoadDatabase()
                     } else {
-                        let userSession = self.startBackgroundSession(for: account, with: coreDataStack)
-
-                        self.migrateCryptoboxIfNeeded(in: coreDataStack.accountContainer, syncContext: userSession.syncContext) {
+                        let userSession = self.startBackgroundSession(
+                            for: account,
+                            with: coreDataStack
+                        )
+
+                        self.migrateCryptoboxSessionsIfNeeded(
+                            in: coreDataStack.accountContainer,
+                            syncContext: userSession.syncContext
+                        ) {
                             completion(userSession)
                         }
 
                     }
+
                     onWorkDone()
                     group?.leave()
                 }
             }
-        })
+        }
     }
 
-    fileprivate func migrateCryptoboxIfNeeded(in accountDirectory: URL,
-                                              syncContext: NSManagedObjectContext,
-                                              completion: @escaping () -> Void) {
-        guard self.cryptoboxMigrationManager.isNeeded(in: accountDirectory) else {
+    /// Migrates all existing proteus data created by Cryptobox into Core Crypto, if needed.
+
+    private func migrateCryptoboxSessionsIfNeeded(
+        in accountDirectory: URL,
+        syncContext: NSManagedObjectContext,
+        completion: @escaping () -> Void
+    ) {
+        guard cryptoboxMigrationManager.isMigrationNeeded(accountDirectory: accountDirectory) else {
+            WireLogger.proteus.info("cryptobox migration is not needed")
+
+            syncContext.performAndWait {
+                do {
+                    try cryptoboxMigrationManager.completeMigration(syncContext: syncContext)
+                } catch {
+                    fatalError("failed to complete proteus initialization")
```

Comment:
so if it fails, we don't have a way to restore the previous state, right?
it will keep on crashing, do we have a recovery plan in this case? What do the other platform do?

Reply:
I suppose this would fail if we fail to create the proteus service, which I would consider a developer error. An acceptable way to handle it could be to log the error and find what in the proteus service setup code can go wrong

Reply:
I think Android crashes if any part of the migration isn't successful.